File: db4o.net/Db4objects.Db4o/Db4objects.Db4o/Internal/LocalPersistentBase.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

using Db4objects.Db4o;
using Db4objects.Db4o.Internal;
using Db4objects.Db4o.Internal.Ids;
using Db4objects.Db4o.Internal.Slots;

namespace Db4objects.Db4o.Internal
{
	/// <exclude></exclude>
	public abstract class LocalPersistentBase : PersistentBase
	{
		private readonly ITransactionalIdSystem _idSystem;

		public LocalPersistentBase(ITransactionalIdSystem idSystem)
		{
			_idSystem = idSystem;
		}

		public LocalPersistentBase() : this(null)
		{
		}

		public override ITransactionalIdSystem IdSystem(Transaction trans)
		{
			if (_idSystem != null)
			{
				return _idSystem;
			}
			return base.IdSystem(trans);
		}

		protected override ByteArrayBuffer ReadBufferById(Transaction trans)
		{
			Slot slot = IdSystem(trans).CurrentSlot(GetID());
			if (DTrace.enabled)
			{
				DTrace.SlotRead.LogLength(GetID(), slot);
			}
			return ((LocalObjectContainer)trans.Container()).ReadBufferBySlot(slot);
		}
	}
}

File: db4oj/core/src/com/db4o/internal/LocalPersistentBase.java
/* Copyright (C) 2004 - 2010  Versant Inc.  http://www.db4o.com */

package com.db4o.internal;

import com.db4o.*;
import com.db4o.internal.ids.*;
import com.db4o.internal.slots.*;

/**
 * @exclude
 */
public abstract class LocalPersistentBase extends PersistentBase {
	
	private final TransactionalIdSystem _idSystem;
	
	public LocalPersistentBase(TransactionalIdSystem idSystem){
		_idSystem = idSystem;
	}
	
	public LocalPersistentBase(){
		this(null);
	}
	
	public TransactionalIdSystem idSystem(Transaction trans) {
		if(_idSystem != null){
			return _idSystem;
		}
		return super.idSystem(trans);
	}
	
	@Override
	protected ByteArrayBuffer readBufferById(Transaction trans) {
		Slot slot = idSystem(trans).currentSlot(getID());
		if(DTrace.enabled){
			DTrace.SLOT_READ.logLength(getID(), slot);
		}
		return ((LocalObjectContainer)trans.container()).readBufferBySlot(slot);
	}

}
